Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+        switch (updateItem.updateItemType) {
+            case ConversationUpdateItemType_Delete:
+                isOnlyModifyingLastMessage = NO;
+                break;
+            case ConversationUpdateItemType_Insert: {
+                id<ConversationViewItem> viewItem = updateItem.viewItem;
+                OWSAssertDebug(viewItem);
+                if (([viewItem.interaction isKindOfClass:[TSIncomingMessage class]] ||
+                        [viewItem.interaction isKindOfClass:[TSOutgoingMessage class]])
+                    && updateItem.newIndex >= oldViewItemCount) {
+                    continue;
+                }
+                isOnlyModifyingLastMessage = NO;
+                break;
+            }
+            case ConversationUpdateItemType_Update: {
+                id<ConversationViewItem> viewItem = updateItem.viewItem;
+                if ([updatedNeighborItemSet containsObject:viewItem.itemId]) {
+                    continue;
+                }
+                OWSAssertDebug(viewItem);
+                if (([viewItem.interaction isKindOfClass:[TSIncomingMessage class]] ||
+                        [viewItem.interaction isKindOfClass:[TSOutgoingMessage class]])
+                    && updateItem.newIndex >= oldViewItemCount) {
+                    continue;
+                }
+                isOnlyModifyingLastMessage = NO;
+                break;
+            }
+        }
+    }
+    BOOL shouldAnimateRowUpdates = !isOnlyModifyingLastMessage;
+    return shouldAnimateRowUpdates;
+}
+
+- (void)createNewMessageMappings
+{
+    if (self.thread.uniqueId.length > 0) {
+        self.messageMappings = [[YapDatabaseViewMappings alloc] initWithGroups:@[ self.thread.uniqueId ]
+                                                                          view:TSMessageDatabaseViewExtensionName];
+    } else {
+        OWSFailDebug(@"uniqueId unexpectedly empty for thread: %@", self.thread);
+        self.messageMappings =
+            [[YapDatabaseViewMappings alloc] initWithGroups:@[] view:TSMessageDatabaseViewExtensionName];
+    }
+
+    // Cells' appearance can depend on adjacent cells in both directions.
+    //
+    // TODO: We could do the same thing in our logic to generate "update items"
+    //       in updateViewWitholdItemIdList.
+    [self.messageMappings setCellDrawingDependencyOffsets:[NSSet setWithArray:@[
+        @(-1),
+        @(+1),
+    ]]
+                                                 forGroup:self.thread.uniqueId];
+
+    [self.uiDatabaseConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+        [self.messageMappings updateWithTransaction:transaction];
+    }];
+    // We need to impose the range restrictions on the mappings immediately to avoid
+    // doing a great deal of unnecessary work and causing a perf hotspot.
+    [self updateMessageMappingRangeOptions];
+}
+
+- (void)resetMappings
+{
+    // If we're entering "active" mode (e.g. view is visible and app is in foreground),
```

Comment:
I don't think this comment is relevant any longer - we no longer stop observing, right?

Reply:
I'm assuming everything after here was moved unchanged from the CVC, LMK if there's something I should look at.

Reply:
> I don't think this comment is relevant any longer - we no longer stop observing, right?

Removed.

> I'm assuming everything after here was moved unchanged from the CVC, LMK if there's something I should look at.

`ensureLoadWindowContainsQuotedReply` is a new method, but it was derived from similar logic in CVC.